### ISSUE TITLE
training model is not stable

### ISSUE THREAD

**Issue by benwu232** [dmlc/xgboost#1838]
Hi guys,

I'm using sklean wrapper. After training, I save the model buy pickle. Next time, I load the pickle and use the model to predict. I found the training model is not stable. Each time after loading a model by pickle, I  run same test with same data but get different results. Sometimes the differences of the results are not ignorable. However, after loading the model (it is in memory), when I run same model with same data, I get same results. It seems every time it is loaded (by pickle), the model changes a little bit. I tried load a model and save it to another file, then compared them. They are not equal. I'm not sure if it is wrong or because of pickle format.

I checked the issues and found some similar ones.
#13
#1205

Here is @tqchen 's answer in #13:
Seems the problem is floating point execution and since we are doing multi-threading here, different summation order of float causes the accuracy difference, which results in difference in latter trees.

I tried to change the nthread to 1 and run my test case, it still got different results (in each loading model buy pickle). Now I doubt it is because of random order of parallel excutions.  

Has anybody seen similar problem and got any solution? 

Thanks a lot!

Here is my environment:
Ubuntu 14.04/64-bit
Python 3.5.2 on Anaconda
xgboost 0.6a2 pip install version

**Comment by benwu232** [comment]
Clarification:
1. Each time I load the same model and run it with same test data, I get different result.
2. With loaded model, each time I run it with same test data, I get same result.
It seems some hidden thing is done during loading a model, which makes same model cause different results. I'm using pickle so far.

**Comment by benwu232** [comment]
Here is an example of my model by pickle. Each time I load it, the result is different.

For example:
a = pickle.load(open('mymodel', 'rb'))
b = pickle.load(open('mymodel', 'rb'))
print(a == b)

I got result False.

Is there anything wrong with pickle? Can we privide a saving method for sklearn wrapper?

Here is the link of the sample model:
https://pan.baidu.com/s/1nvrZ0Wp

Many thanks!

**Comment by benwu232** [comment]
Finally I found the problem, which is nothing to do with xgboost or pickle. There is something else which causes some randomization in each running and makes the result different.

Thanks for help! Sorry for bothering!

**Comment by sjayrams** [comment]
@benwu232 was it in sklearn?

**Comment by benwu232** [comment]
@sjayrams  It's in my code, not sklearn.

**Comment by N3da** [comment]
@benwu232 Just curious, could you explain what the issue was that didn't repro when the model was already loaded into memory but only when it was reloaded?
